File: src/kyu_6/prize_draw/js/rank.js
function score(s) {
    let alph = "abcdefghijklmnopqrstuvwxyz";
    let sum = 0;
    for (let el of s) {
        sum += alph.indexOf(el) + 1;
    }
    return sum + s.length;
}

function rank(st, we, n) {
    if (st === "") {
        return "No participants"
    }
    let s = st.split(',')
    if (n > s.length) {
        return "Not enough participants"
    }
    let st1 = st.toLowerCase().split(",");
    let w = st1.map((el, i) => [s[i], score(el) * we[i]])
    w.sort((a, b) => b[1] > a[1] ? 1 : b[1] < a[1] ? -1 : a[0] > b[0] ? 1 : -1)
    return w[n - 1][0]
}

module.exports = rank;

File: src/kyu_6/prize_draw/java/Rank.java
package kyu_6.prize_draw.java;

import java.util.HashMap;
import java.util.Map;

class Rank {

    public static String nthRank(String st, Integer[] we, int n) {
        if (st.isEmpty()) {
            return "No participants";
        } else if(n > we.length) {
            return "Not enough participants";
        }
        String[] names = st.split(",");
        Map<String, Integer> ranks = new HashMap<>();
        for (int i = 0; i < names.length; i++) {
            String name = names[i];
            int weight = we[i];
            int rank = (name.length() + name.chars().map(c -> Character.toLowerCase(c) - 'a' + 1).sum()) * weight;
            ranks.put(name, rank);
        }
        return ranks.entrySet()
                .stream()
                .sorted(Map.Entry.<String, Integer>comparingByValue().reversed()
                        .thenComparing(Map.Entry.comparingByKey()))
                .map(Map.Entry::getKey)
                .toArray(String[]::new)[n - 1];
    }
}